Better string representation of errors

macjohnny opened the pull request:
Resolves #2711

Before:
```
Chrome 59.0.3071 (Mac OS X 10.12.3) ERROR
 {
      "originalErr": {},
      "__zone_symbol__currentTask": {
        "type": "microTask",
        "state": "notScheduled",
        "source": "Promise.then",
        "zone": "<root>",
        "cancelFn": null,
        "runCount": 0
      }
  }
Chrome 59.0.3071 (Mac OS X 10.12.3): Executed 0 of 0 ERROR (1.348 secs / 0 secs)
```

After:
```
Chrome 59.0.3071 (Mac OS X 10.12.3) ERROR
 {
    "message": {
      "originalErr": {},
      "__zone_symbol__currentTask": {
        "type": "microTask",
        "state": "notScheduled",
        "source": "Promise.then",
        "zone": "<root>",
        "cancelFn": null,
        "runCount": 0
      }
    },
    "str": [
      "Error: (SystemJS) Object prototype may only be an Object or null: undefined",
      "    TypeError: Object prototype may only be an Object or null: undefined",
      "        at setPrototypeOf (<anonymous>)",
      "        at __extends$1 (dist/bundles/my-bundle.umd.js:2258:9)",
    ]
  }
Chrome 59.0.3071 (Mac OS X 10.12.3): Executed 0 of 0 ERROR (1.348 secs / 0 secs)
```

macjohnny commented:
I signed it!

wesleycho commented on client/karma.js:
```diff
@@ -112,7 +112,12 @@ var Karma = function (socket, iframe, opener, navigator, location) {
+    
+    // create an object with the string representation of the message to ensure all its content is properly 
+    // transferred to the console log
+    message = {message: message, str: message.toString().replace(/\t/g, "    ").split("\n")}
```
The `replace` and `split` parts should probably be removed.

Also, this line changes the message variable's type from string to an object - this probably means that tests have to be updated.

macjohnny replied:
the replace and split parts are for nicer output display in the console, since the object is converted to json. 
With the split and replace, you get 
```
"str": [
      "Error: (SystemJS) Object prototype may only be an Object or null: undefined",
      "    TypeError: Object prototype may only be an Object or null: undefined",
      "        at setPrototypeOf (<anonymous>)",
      "        at __extends$1 (dist/bundles/my-bundle.umd.js:2258:9)",
    ]
```
instead of 

```
"str": "Error: (SystemJS) Object prototype may only be an Object or null: undefined\n\tTypeError: Object prototype may only be an Object or null: undefined\n\tat setPrototypeOf (<anonymous>)\n\tat __extends$1 (dist/bundles/my-bundle.umd.js:2258:9)",
```

macjohnny replied:
the variable type is not changed, the message already is an object that will later on be converted to a string with JSON.stringify()

macjohnny replied:
the message might be a string or an object

macjohnny commented:
@wesleycho could you have a look at this again? this patch really helped me a lot when investigating preliminary import / systemjs issues that prevented the tests from being run and log a useful message.

macjohnny commented:
@wesleycho could you please have a look at this again?

larsduelfer commented:
I would also highly appreciate if this change is going to be integrated.

macjohnny commented:
@EzraBrooks I can understand the formatting should not be done here. but since the message *can* be an object (which is the reason for this PR), it should be ok to transform it to a different object as I suggest in this PR.  would you accept this PR if I removed the `.replace(/\t/g, "    ").split("\n")` part?

EzraBrooks commented:
Thanks, I’ll approve it once the CI checks complete.

macjohnny commented:
@EzraBrooks it seems there is a problem with the travis ci build retrieving some commit message. however, all other tests pass. could you please have a look at it?

EzraBrooks commented:
Please squash your commits into one that follows Karma’s naming conventions - Travis is failing on commit message validation.

EzraBrooks commented:
Sorry, to clarify.. the commit messages all have a leading type on them. In this case, I would recommend `feat: better string representation of errors`.

macjohnny commented:
@EzraBrooks done, waiting for the CI to complete

EzraBrooks commented:
Looks like ESLint has a few style complaints.

macjohnny commented:
@EzraBrooks I updated the PR, hope it is fine now
